Update BrowserStack browser testing docs (#36269)
mdn/content: files/en-us/learn/tools_and_testing/cross_browser_testing/automated_testing/index.md

   npm -v
   ```

> 4. If you've got Node/npm already installed, or installed them a while ago, you should update them to their latest versions. To update Node, the most reliable way is to download and install an updated installer package from their website (see link above). To update npm, use the following command in your terminal:

Reviewer: I don't see the value in this addition/feels like pointless duplication. Do you update to the latest version even if you just installed or not? Assuming you don't you would do something like.

```suggestion
4. If you've previously installed Node/npm, you should update them to their latest versions. To update Node, the most reliable way is to download and install an updated installer package from their website (see link above). To update npm, use the following command in your terminal:
```

Note, I won't force this on you, but the best way to install node/nvm is to use a package manager like nvm. We probably should have a single document on that and link. I cover this in https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/development_environment#installing_node

Author: Yup, good call, updated.

I also decided to shorten my section down and just link to yours for the most part — it is much better than mine!

Now let's look at commercial third-party browser testing services and what they can do for us.

> The basic premise with such applications is that the company that runs each one has a huge server farm that can run many different tests. When you use this service, you provide a URL of the page you want to test along with information, such as what browsers you want it tested in. The app then configures a new VM with the OS and browser you specified, and returns the test results in the form of screenshots, videos, log files, text, etc. This is very useful, and way more convenient than having to set up all the OS/browser combinations by yourself.

Reviewer: The first sentence feels awkward. Can e do without it?

```suggestion
When you use these kinds of services, you provide a URL of the page you want to test along with information, such as what browsers you want it tested in. The app then configures a new VM with the OS and browser you specified, and returns the test results in the form of screenshots, videos, log files, text, etc. This is very useful, and way more convenient than having to set up all the OS/browser combinations by yourself.
```

Author: Yes, I agree with that. Updated.

You can then step up a gear, using an API to access functionality programmatically, which means that such apps can be combined with task runners, such as your own local Selenium environments and others, to create automated tests.

> [!NOTE]

> > There are other commercial browser testing systems available but in this article, we'll focus on Sauce Labs, BrowserStack, and TestingBot. We're not saying that these are necessarily the best tools available, but they are good ones that are simple for beginners to get up and running with.

Reviewer: It's a pity you've dropped lambdatest - "Unlike other cloud-based cross browser testing service providers, LambdaTest offers a freemium account where you get lifetime access to their platform."

It is very "MDN" to over at least one option that is free and accessible to all. Appreciate the reasons. Do the others offer free options? Might be worth outlining (if you haven't - I'm working my way down)

Author: Thing is, I looked at LambdaTest and what it offers today; they used to offer more free minutes and stuff, but these days the free offering didn't seem much better than what the others offer. They all have a free offering of sorts, but the general case is that you get a really short test time (like, a minute or two per session).

   ```

3. Replace the placeholders for BrowserStack username and access key with your actual values. These can be retrieved from your [BrowserStack Account & Profile Details](https://www.browserstack.com/accounts/profile/details), under the _Authentication & Security_ section.

> 4. Install the [axios](https://www.npmjs.com/package/axios) module we are using in the code to handle sending HTTP requests by running the following command in your terminal (we chose axios because it is simple, popular, and well-supported):

Reviewer: Well so is Request, but I'm not arguing :-)

Author: Request is popular, but it has also been deprecated; read https://github.com/request/request/issues/3142. I therefore decided to change to a library that isn't, for better longevity.

I also like the axois syntax better — feels cleaner and a bit more modern.

Author: But I didn't think it was helpful to the target audience to start going into all this.

Reviewer: Thanks for the explanation. Fair enough.!

You'll also see a menu that allows you to control the session.

![Test Menu](browserstack-test-menu-sized.png)

The available features vary depending on what browser is loaded, and can include controls for:

- Displaying information on the current browser
- Switching to other browsers
- Testing localhost URLs
- Setting zoom level and toggling orientation
- Saving and loading bookmarks
- Capturing/annotating screenshots and filing bug reports
- Accessing browser DevTools
- Changing reported location
- Throttling the network
- Accessing screenreaders

#### Advanced: The BrowserStack API

BrowserStack also has a [restful API](https://www.browserstack.com/docs/automate/api-reference/selenium/introduction) that allows you to programmatically retrieve details of your account plan, sessions, builds, etc.

Let's have a brief look at how we'd access the API using Node.js.

1. First, set up a new npm project to test this out, as detailed in [Setting up Node and npm](#setting_up_node_and_npm). Use a different directory name than before, like `bstack-test` for example.
2. Create a new file inside your project root called `call_bstack.js` and give it the following content:

   ```js

>    const axios = require("axios");

Reviewer: This is new to me. I know you refer to it further down, but you might want to forward-reference it, and explain why you're using it rather than request.

Author: I noticed that I erroneously left in a mention of `request` somewhere. I've removed that now.

I've also added a piece of text to explain why this library was chosen.
